libtransmission/variant.c

```diff
@@ -1141,8 +1141,21 @@ tr_variantMergeDicts (tr_variant * target, const tr_variant * source)
             }
           else if (tr_variantIsList (val))
             {
-              if (tr_variantDictFind (target, key) == NULL)
-                tr_variantListCopy (tr_variantDictAddList (target, key, tr_variantListSize (val)), val);
+              tr_variant * targetList = tr_variantDictFind (target, key);
+              if (targetList == NULL)
+                {
+                  tr_variantListCopy (tr_variantDictAddList (target, key, tr_variantListSize (val)), val);
+                }
+              else
+                {
+                  // if we are processing TR_KEY_download_groups list
+                  // overwrite target with source
+                  if (key == TR_KEY_download_groups && tr_variantListSize(val) > 0)
```

Comment:
variant.c `tr_variantMergeDicts()` is refactor of libtransmission's tr_benc class as tr_variant from bencode.c `tr_bencMergeDicts()` (2.8x)

variant.c `tr_variantMergeDicts()` line 1142

```
else if (tr_variantIsList (val))
  {
    if (tr_variantDictFind (target, key) == NULL)
      tr_variantListCopy (tr_variantDictAddList (target, key, tr_variantListSize (val)), val);
  }
```


was derived from https://github.com/transmission/transmission/commit/36d97610bde962c10bab53d627015802c37ce1fe#diff-bde72b74342bfa7f9cbaa1fd2bbcb855 (bencode.c  line **1562**) so the function _MergeDicts_ then wasn't all inclusive.

https://trac.transmissionbt.com/ticket/4687#comment:14

> 
> ...
> cfpp2p's patch solves this use case nicely by always adding the **nested** dict whenever merging two dictionaries together and the source has a nested dictionary itself. However, tr_bencDictMerge() is meant to be a general purpose function, and it's not clear to me that is the correct behavior every time. For example, what if the target dictionary has its own nested dictionary with the same key? Ideally we would merge those two children together, rather than one overwriting the other.
> 
> So, here's my variation on cfpp2p's fix:
> 
>     else if( tr_bencIsDict( val ) )
>     {
>         tr_benc * target_dict = tr_bencDictFind( target, key );
> 
>         if( target_dict == NULL )
>             target_dict = tr_bencDictAddDict( target, key, tr_bencDictSize( val ) );
> 
>         if( tr_bencIsDict( target_dict ) )
>             tr_bencMergeDicts( target_dict, val );
>     }
> 
> This way, we'll Do the Right Thing if the target has no child with that name, or if the child is a dictionary.
> ...
> 
This commit
https://github.com/transmission/transmission/commit/c56b12690c49a9f7c76c20b4f6046002d8e7b228

Since the `download-groups` are nested lists within a list and for sticking with   **_"is meant to be a general purpose function"_**    a similar recursive tactic as https://github.com/transmission/transmission/blob/master/libtransmission/variant.c#L1147
might be one way to proceed.